### PR TITLE
[FLINK-28803] Add Transformer and Estimator for KBinsDiscretizer

### PR DESCRIPTION
**Pull request by zhipeng93**
## What is the purpose of the change
- Add Transformer and Estimator for KBinsDiscretizer[1] in Flink ML. 

## Brief change log
- Added Transformer and Estimator for KBinsDiscretizer.
- Added java test/example  for Transformer and Estimator for KBinsDiscretizer.
- Added python source/test/example for KBinsDiscretizer.
- Comparing with Sklearn [1], we made the following changes:
     - We remove `encode` parameter since we can achieve the same goal with a onehot-encoder following KBinsDiscretizer.
     - We remove `random_state` since it is non-trivial to support reproduciable sampling in a distributed setting, while in sklearn (single-node setting), it is much easier. If the users ask for this feature, we can add it later.
     - We remove `dtype` since we can achieve the same goal with a simple type conversion following KBinsDiscretizer.
     - We support `sampling` for both `quantile` and `kmeans`, but sklearn supports `sampling` only for `quantile`. [2]
     - We hardcode the threshold for width of bins as zero, while Sklearn uses 1e-8. [3]
     - We do not support `numSamples` as None while sklearn supports. [4]

## Does this pull request potentially affect one of the following parts:
- Dependencies (does it add or upgrade a dependency): (no)
- The public API, i.e., is any changed class annotated with @Public(Evolving): (no)
- Does this pull request introduce a new feature? (yes)
- If yes, how is the feature documented? (Java doc)

[1] https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.KBinsDiscretizer.html
[2] https://github.com/apache/flink-ml/pull/139#discussion_r943058613
[3] https://github.com/apache/flink-ml/pull/139#discussion_r943052952
[4] https://github.com/apache/flink-ml/pull/139#discussion_r943053798

### CODE REVIEW

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+/**
+ * An Estimator which implements discretization (also known as quantization or binning), which
+ * transforms continuous features into discrete ones. The output values are in [0, numBins).
+ *
+ * <p>KBinsDiscretizer implements three different binning strategies, and it can be set by {@link
+ * KBinsDiscretizerParams#STRATEGY}. If the strategy is set as {@link KBinsDiscretizerParams#KMEANS}
+ * or {@link KBinsDiscretizerParams#QUANTILE}, users should further set {@link
+ * KBinsDiscretizerParams#SUB_SAMPLES} for better performance.
+ *
+ * <p>There are several cornel cases for different inputs as listed below:
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:60]: nit: "corner"

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+ *
+ * <ul>
+ *   <li>When the input values of one column are all the same, then they should be mapped to the
+ *       same bin (i.e., the zero-th bin). Thus the corresponding bin edges are {Double.MIN_VALUE,
+ *       Double.MAX_VALUE}.
+ *   <li>When the number of distinct values of one column is less than the specified number of bins
+ *       and the {@link KBinsDiscretizerParams#STRATEGY} is set as {@link
+ *       KBinsDiscretizerParams#KMEANS}, we switch to {@link KBinsDiscretizerParams#UNIFORM}.
+ *   <li>When the width of one output bin is zero, i.e., the left edge equals to the right edge of
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:69]: SKlearn removes bins whose width is less than `1e-8`[1]. Shall we adopt this value as well, or introduce a `HasTol` parameter to allow users to specify the width of the bins to be ignored?

[1] https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.KBinsDiscretizer.html

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    Param<Integer> NUM_BINS =
+            new IntParam("numBins", "Number of bins to produce.", 5, ParamValidators.gtEq(2));
+
+    Param<Integer> SUB_SAMPLES =
+            new IntParam(
+                    "subSamples",
+                    "Maximum number of samples used to fit the model.",
+                    200000,
+                    ParamValidators.gt(0));
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:60]: sklearn allows this value to be set to `None`, in which case all the training samples are used when computing the quantiles that determine the binning thresholds. Shall we support this value as well?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import org.apache.flink.ml.param.Param;
+import org.apache.flink.ml.param.ParamValidators;
+import org.apache.flink.ml.param.StringParam;
+
+/**
+ * Params for {@link KBinsDiscretizer}.
+ *
+ * @param <T> The class type of this instance.
+ */
+public interface KBinsDiscretizerParams<T> extends KBinsDiscretizerModelParams<T> {
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:31]: Would it be better to inherit `HasSeed` in this class to provide the random seed for sub sampling?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerModel.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A Model which transforms continuous features into discrete features using the model data computed
+ * by {@link KBinsDiscretizer}.
+ *
+ * <p>A feature value {v} should be mapped to a bin with edges as {left, right} if {v} is in [left,
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerModel.java:51]: Is usages like `{v}` a common way to represent variables? I have not seen other usages like this before.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+                                    "reduceInEachPartition",
+                                    inputData.getType(),
+                                    new MinMaxReduceFunctionOperator())
+                            .transform(
+                                    "reduceInFinalPartition",
+                                    inputData.getType(),
+                                    new MinMaxReduceFunctionOperator())
+                            .setParallelism(1);
+        } else {
+            preprocessedData = DataStreamUtils.sample(inputData, getSubSamples(), 2022L);
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:113]: According to sklearn's document, the subsample parameter is only used when `strategy="quantile"`[1]. Here, it seems that the sampling is also applied when `strategy="kmeans"`.

[1] https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.KBinsDiscretizer.html

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+
+            if (features[0] == features[numData - 1]) {
+                LOG.warn("Feature " + columnId + " is constant and the output will all be zero.");
+                binEdges[columnId] = new double[] {Double.MIN_VALUE, Double.MAX_VALUE};
+            } else {
+                // Checks whether there are more than {numBins} distinct feature values in each
+                // column.
+                // If the number of distinct values is less than {numBins + 1}, then we do not need
+                // to conduct KMeans. Instead, we switch to use {@link
+                // KBinsDiscretizerParams#UNIFORM} for bining.
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:263]: nit: "binning".

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+                    // Conducts KMeans here.
+                    double width = 1.0 * features.length / numBins;
+                    for (int clusterId = 0; clusterId < numBins; clusterId++) {
+                        kMeansCentroids[clusterId] = features[(int) (clusterId * width)];
+                    }
+
+                    // Default values for KMeans following sklearn:
+                    // https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.KBinsDiscretizer.html#sklearn-preprocessing-kbinsdiscretizer.
+                    final double tolerance = 1e-4;
+                    final int maxIterations = 300;
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:291]: I was unable to find information like `maxIterations = 300` in the link provided above. Do you know where I can find them?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+                    double max = features[features.length - 1];
+                    double width = (max - min) / numBins;
+                    binEdges[columnId] = new double[numBins + 1];
+                    binEdges[columnId][0] = min;
+                    for (int edgeId = 1; edgeId < numBins + 1; edgeId++) {
+                        binEdges[columnId][edgeId] = binEdges[columnId][edgeId - 1] + width;
+                    }
+                    continue;
+                } else {
+                    // Conducts KMeans here.
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:282]: Would it be better to reuse the logic in the kmeans or online kmeans stages?

**Review by yunfengzhou-hub — COMMENTED** [review]
Thanks for the PR. Left some comments as below.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java (hunk 2 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:69]: Removing bins whose width is less than `1e-8` is ambiguous to users --- Consider the following case:
Suppose we have four consecutive bins with width {6e-9, 6e-9, 6e-9, 6e-9} and the minimum bin width is `1e-8`, we may get different results: {1.2e-8, 1.2e-8} and {2.4e-8}, both of which are reasonable.

Setting minimum bin width as zero does not incur the above problem. If the users found the bin is too small, they may decrease number of bins.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java (hunk 3 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:60]: Setting the value as `None` is useless here. If users want to use all of the samples, they can set the value as `Integer.MAX_VALUE`.

In our implementation, if `numSamples` is greater than number of input data, all of the data are used in quantile/kmeans.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java (hunk 4 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:31]: Setting `HasSeed` has two possible indications:
- Using same seed can have reproduciable outputs
- Using different seed can have different outputs.

However, our implementation does not provide reproduciable outputs due to the complexity of reproduciable distributed sampling --- We always provide different outputs in each run. As a result, it seems meaningless to provide `HasSeed` here.

However, we can add `HasSeed` later if we can provide reproduciable distributed sampling.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java (hunk 8 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:291]: It uses the default setting in Kmeans [1][2].

[1] https://github.com/scikit-learn/scikit-learn/blob/17df37aee774720212c27dbc34e6f1feef0e2482/sklearn/preprocessing/_discretization.py#L279
[2] https://github.com/scikit-learn/scikit-learn/blob/17df37aee774720212c27dbc34e6f1feef0e2482/sklearn/cluster/_kmeans.py#L1162

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java (hunk 6 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:113]: As I understand, `quantitle` and `kmeans` are similar in the following aspects:
- They are both used to estimate/compute the bin edges.
- For accurate implementation, they both need to access/store all of the input data.

Now that sampling is used for `quantitle` to accelerate the process of computing bins, with a compromise on precision. Then it seems unclear why do we do different stuff to `kmeans`. 

If we do not support sampling for `kmeans`, it may only work on small datasets. Also, users can also set `numSamples` as `Integer.MAX_VALUE` to do `kmeans` with no sampling.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java (hunk 9 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:282]: It may no be that efficient to reuse the logic in `KMeans` class:
- We do kmeans for each input column here, i.e., the input feature is simply a double, not a vector.
- Reusing `Kmeans` class means we introduce extra flink operators, and make the job more complex. Here a toy code for one-dimension kmeans is enough.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerModel.java (hunk 5 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerModel.java:51]: Thanks for pointing this out. I have replaced `{}` with `\`\`` following [1]

[1] https://github.com/apache/spark/blob/e17d8ecabcad6e84428752b977120ff355a4007a/mllib/src/main/scala/org/apac

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java (hunk 4 above)
Comment (yunfengzhou-hub) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:31]: Got it. I agree with it that we may try to support this parameter in future if we need it.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java (hunk 3 above)
Comment (yunfengzhou-hub) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:60]: Got it. I agree with it that we may try to support this parameter in the future if we need it.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java (hunk 2 above)
Comment (yunfengzhou-hub) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:69]: Got it. I agree with it that we may try to support this parameter in the future if we need it.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java (hunk 6 above)
Comment (yunfengzhou-hub) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:113]: Sounds good.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java (hunk 3 above)
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:60]: nit: Shall we add a check that `subSamples >= 2` or `subSamples >= numBins`?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+                    double width = 1.0 * features.length / numBins;
+                    for (int clusterId = 0; clusterId < numBins; clusterId++) {
+                        kMeansCentroids[clusterId] = features[(int) (clusterId * width)];
+                    }
+
+                    // Default values for KMeans following sklearn:
+                    // https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.KBinsDiscretizer.html#sklearn-preprocessing-kbinsdiscretizer.
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:289]: Do you think it would be better to remove comments like this? This information is useful for reviewers, that the implementation is following an existing practice. But I wonder if this information will also be necessary for future readers, or if this information will be outdated when sklearn got updated.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+        env = StreamExecutionEnvironment.getExecutionEnvironment(config);
+        env.setParallelism(4);
+        env.enableCheckpointing(100);
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        tEnv = StreamTableEnvironment.create(env);
+        trainTable = tEnv.fromDataStream(env.fromCollection(TRAIN_INPUT)).as("input");
+        testTable = tEnv.fromDataStream(env.fromCollection(TEST_INPUT)).as("input");
+    }
+
+    @SuppressWarnings("all")
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java:138]: nit: it might be better to replace this with the detailed warnings you want to suppress.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    private void verifyPredictionResult(
+            List<Row> expectedOutput, Table output, String predictionCol) throws Exception {
+        List<Row> collectedResult =
+                IteratorUtils.toList(
+                        tEnv.toDataStream(output.select($(predictionCol))).executeAndCollect());
+        compareResultCollections(
+                expectedOutput,
+                collectedResult,
+                (o1, o2) -> {
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java:147]: There is a method called `TestUtils::compare`, which might be useful here. And #142 has been generalizing the method from `DenseVector` to `Vector`.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+
+    @Test
+    public void testFitOnEmptyData() {
+        Table emptyTable =
+                tEnv.fromDataStream(env.fromCollection(TRAIN_INPUT).filter(x -> x.getArity() == 0))
+                        .as("input");
+        KBinsDiscretizerModel model = new KBinsDiscretizer().fit(emptyTable);
+        Table modelDataTable = model.getModelData()[0];
+        try {
+            IteratorUtils.toList(
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java:288]: nit: `modelDataTable.execute().collect().next();` might be simpler here.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+                                    "reduceInEachPartition",
+                                    inputData.getType(),
+                                    new MinMaxReduceFunctionOperator())
+                            .transform(
+                                    "reduceInFinalPartition",
+                                    inputData.getType(),
+                                    new MinMaxReduceFunctionOperator())
+                            .setParallelism(1);
+        } else {
+            preprocessedData = DataStreamUtils.sample(inputData, getSubSamples(), 2022L);
```
Comment (yunfengzhou-hub) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:113]: Would it be better to change the random seed from `2022L` to `getClass().getName().hashCode()`? If one day we would support the random seed parameter, this would be its default value.

**Review by yunfengzhou-hub — COMMENTED** [review]
Thanks for the update. Left some comments as below.

File: flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java (hunk 11 above)
Comment (zhipeng93) [reply flink-ml-lib/src/test/java/org/apache/flink/ml/feature/KBinsDiscretizerTest.java:138]: Thanks for pointing this out. I have updated it with `@SuppressWarnings("unchecked, ConstantConditions")`

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java (hunk 10 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:289]: I do not have a strong preference here. I have removed these comments since you pointed this out.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java (hunk 3 above)
Comment (zhipeng93) [reply flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizerParams.java:60]: I agree that `subSamples < 2` is not a valid use case and I have udpated the code.

If `subSamples < numBins`, it should still work in `quantile` and `kmeans` case.

**Comment by zhipeng93** [comment]
@yunfengzhou-hub Thanks for the review. I have addressed your comments in the PR. 
@lindong28 Do you want to take a look?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 15]
```diff
@@ -0,0 +1,341 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * An Estimator which implements discretization (also known as quantization or binning), which
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:52]: nits: maybe rephrase this sentence like this:

```
 * An Estimator which implements discretization (also known as quantization or binning) to
 * transform continuous features into discrete ones. The output values are in [0, numBins).
```

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 16]
```diff
@@ -0,0 +1,341 @@
...
+            } else {
+                double width = 1.0 * features.length / numBins;
+                double[] tempBinEdges = new double[numBins + 1];
+
+                for (int binEdgeId = 0; binEdgeId < numBins; binEdgeId++) {
+                    tempBinEdges[binEdgeId] = features[(int) (binEdgeId * width)];
+                }
+                tempBinEdges[numBins] = features[numData - 1];
+
+                // Removes bins who are empty, i.e., the left edge equals to the right edge.
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:229]: nits: `who` -> `that`

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java [hunk 17]
```diff
@@ -0,0 +1,341 @@
...
+            Arrays.sort(features);
+
+            if (features[0] == features[numData - 1]) {
+                LOG.warn("Feature " + columnId + " is constant and the output will all be zero.");
+                binEdges[columnId] = new double[] {Double.MIN_VALUE, Double.MAX_VALUE};
+            } else {
+                // Checks whether there are more than {numBins} distinct feature values in each
+                // column.
+                // If the number of distinct values is less than {numBins + 1}, then we do not need
+                // to conduct KMeans. Instead, we switch to use {@link
```
Comment (lindong28) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/kbinsdiscretizer/KBinsDiscretizer.java:264]: nits: `use` -> `using`

**Review by lindong28 — COMMENTED** [review]
Thanks @zhipeng93 for the PR and thanks @yunfengzhou-hub for the review!

The PR looks pretty good. Left just minor comments. @zhipeng93 Please feel free to merge the PR.

**Comment by zhipeng93** [comment]
Thanks for the review @lindong28. Merging now.
